README.txt
EldisArticle 1.0.0

  The IDS web API provides a way for programmers to use the data and metadata contained in the datasets that IDS maintains - namely BRIDGE and ELDIS. The API can be used for search or for retreiving all the metadata about an object stored in the IDS datasets.

  <b>Overview:</b> The Eldis search API is a HTTP service in a REST style architecture. It allows search of datacore assets based upon a query and a number of meta-data criteria. It also allows a searching of the category hierarchy within the datacore. Data is returned as either JSON or XML, and asset records can be returned in a number of formats, including short form and full form.

  Authentication In order to successfully use the API, users are required to sign up with a Unique ID. It is required that these tokens are sent with every request within the HTTP header. To register for an API key visit http://api.ids.ac.uk/accounts/register/
Rate limiting Users are currently limited to 30 HTTP requests per minute. Once this has been exceeded, users will receive an HTTP 400 response from the server.

  Searching This allows the assets in the datacore to be searched. In order to search, users will need to specify the /search/ parameter after the base URL, and follow this with the query for that search (for example, to search for biofuels, /search/biofuels/etc..). Users are further able to filter the results by specifying a combination of the meta-data criteria below. Each of these should appear in the same /key/value/ format.

  For more API details please visit http://api.ids.ac.uk/docs

  This product is covered by a license. See LICENSE.txt

Installation:

 1. Zope Product

    Extract this tarball into your Zope Product folder
    and restart Zope when you did so.

Created by:

    <p><a href=mailto:subhendu.giri@oneworld.net> Subhendu Kumar Giri </a></p>


ChangeLog.txt
EldisArticle 1.0.0 
	- Initial release by OneWorld



__init__.py
# -*- coding: utf-8 -*-
#
# File: Eldis.py
#
#
# GNU General Public License (GPL)
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
# 02110-1301, USA.
#
__version__ = '$Id: __init__.py 159 2012-05-29 08:26:46Z subhendu $'
__author__ = """Subhendu Kumar Giri <subhendu.giri@oneworld.net>"""
__docformat__ = 'plaintext'


# There are three ways to inject custom code here:
#
#   - To set global configuration variables, create a file AppConfig.py.
#       This will be imported in config.py, which in turn is imported in
#       each generated class and in this file.
#   - To perform custom initialisation after types have been registered,
#       use the protected code section at the bottom of initialize().

import logging
logger = logging.getLogger('Eldis')
logger.debug('Installing Product')

import os
import os.path
from Globals import package_home
import Products.CMFPlone.interfaces
from Products.Archetypes import listTypes
from Products.Archetypes.atapi import *
from Products.Archetypes.utils import capitalize
from Products.CMFCore import DirectoryView
from Products.CMFCore import permissions as cmfpermissions
from Products.CMFCore import utils as cmfutils
from Products.CMFPlone.utils import ToolInit
from config import *

DirectoryView.registerDirectory('skins', product_globals)


def initialize(context):
    """initialize product (called by zope)"""
    ##code-section custom-init-top #fill in your manual code here
    ##/code-section custom-init-top

    # imports packages and types for registration

    import EldisArticle

    from AccessControl import allow_module
    allow_module('Products.Eldis.Extensions.get_eldis_rss')

    #from AccessControl import ModuleSecurityInfo	
    #ModuleSecurityInfo('from Products.Eldis.portal').declarePublic('getEldis_states')	

    # Initialize portal content
    all_content_types, all_constructors, all_ftis = process_types(
        listTypes(PROJECTNAME),
        PROJECTNAME)

    cmfutils.ContentInit(
        PROJECTNAME + ' Content',
        content_types      = all_content_types,
        permission         = DEFAULT_ADD_CONTENT_PERMISSION,
        extra_constructors = all_constructors,
        fti                = all_ftis,
        ).initialize(context)

    # Give it some extra permissions to control them on a per class limit
    for i in range(0,len(all_content_types)):
        klassname=all_content_types[i].__name__
        if not klassname in ADD_CONTENT_PERMISSIONS:
            continue

        context.registerClass(meta_type   = all_ftis[i]['meta_type'],
                              constructors= (all_constructors[i],),
                              permission  = ADD_CONTENT_PERMISSIONS[klassname])



EldisArticle.py
# -*- coding: utf-8 -*-
#
# File: EldisArticle.py
#
#
# GNU General Public License (GPL)
#
__version__ = '$Id: EldisArticle.py 159 2012-05-29 08:26:46Z subhendu $'
__author__ = """Subhendu Kumar Giri <subhendu.giri@oneworld.net>"""
__docformat__ = 'plaintext'

from AccessControl import ClassSecurityInfo
from Products.Archetypes.atapi import *
from Products.kupu.plone.ReftextField import ReftextField
from zope.interface import implements
import interfaces

from Products.CMFDynamicViewFTI.browserdefault import BrowserDefaultMixin

from Products.Eldis.config import *

schema = Schema((

    ReftextField(
        name='bodyText',
        widget=RichWidget(
            label='Bodytext',
            label_msgid='Eldis_label_bodyText',
            i18n_domain='Eldis',
            ),
        ),

    TextField(
        name='sourceUrl',
        widget=StringWidget(
            label='Eldis Article: Source URL',
            description='Auto generated',
            label_msgid='Eldis_label_sourceUrl',
            i18n_domain='Eldis',									
            size = 40,
            ),
        ),		

    TextField(
        name='sourceId',
        index='FieldIndex:schema',
        widget=StringWidget(
            label='Article: Source Unique Id',
            description='Auto generated',
            label_msgid='Eldis_label_sourceId',
            i18n_domain='Eldis',
            size = 40,
            ),
        ),

),
)

EldisArticle_schema = BaseSchema.copy() + \
    schema.copy()


class EldisArticle(BaseContent, BrowserDefaultMixin):
    """
    """
    security = ClassSecurityInfo()

    implements(interfaces.IEldisArticle)

    meta_type = 'EldisArticle'

    _at_rename_after_creation = True

    schema = EldisArticle_schema

 
registerType(EldisArticle, PROJECTNAME)


setuphandlers.py
# -*- coding: utf-8 -*-
#
# File: setuphandlers.py
#
#
# GNU General Public License (GPL)
#
__version__ = '$Id: setuphandlers.py 159 2012-05-29 08:26:46Z subhendu $'
__author__ = """Subhendu Kumar Giri <subhendu.giri@oneworld.net>"""
__docformat__ = 'plaintext'


import logging
logger = logging.getLogger('Eldis: setuphandlers')
from Products.Eldis.config import PROJECTNAME
from Products.Eldis.config import DEPENDENCIES
import os
from Products.CMFCore.utils import getToolByName
import transaction
##code-section HEAD
##/code-section HEAD

def isNotEldisProfile(context):
    return context.readDataFile("Eldis_marker.txt") is None



def updateRoleMappings(context):
    """after workflow changed update the roles mapping. this is like pressing
    the button 'Update Security Setting' and portal_workflow"""
    if isNotEldisProfile(context): return 
    wft = getToolByName(context.getSite(), 'portal_workflow')
    wft.updateRoleMappings()


def postInstall(context):
    """Called as at the end of the setup process. """
    # the right place for your custom code
    if isNotEldisProfile(context): return 
    site = context.getSite()


##code-section FOOT
##/code-section FOOT


Extensions/Install.py
# -*- coding: utf-8 -*-
#
# File: Install.py
#
#
__version__ = '$Id: Install.py 159 2012-05-29 08:26:46Z subhendu $'
__author__ = """Subhendu Kumar Giri <subhendu.giri@oneworld.net>"""
__docformat__ = 'plaintext'


from StringIO import StringIO
from Products.CMFCore.utils import getToolByName
from Products.Eldis.config import PROJECTNAME, PROPERTIES



# Configlets to be added to control panels or removed from them
configlets = (
		{ 'id'         : PROJECTNAME
		, 'name'       : 'Eldis Configuration'
		, 'action'     : 'string:${portal_url}/prefs_eldis_form'
		, 'condition'  : ''
		, 'category'   : 'Products'   # section to which the configlet should be added:
																	# (Plone,Products,Members)
		, 'visible'    : 1
		, 'appId'      : 'EldisMenuSection'
		, 'permission' : 'Manage portal'
		, 'imageUrl'   : 'eldis_icon.gif'
		},
)

def install(self, reinstall=False):
		"""External Method to install Eldis 

		This method to install a product is kept, until something better will get
		part of Plones front end, which utilize portal_setup.
		"""
		out = StringIO()
		print >> out, "Installation log of %s:" % PROJECTNAME

		setuptool = getToolByName(self, 'portal_setup')
		importcontext = 'profile-Products.%s:default' % PROJECTNAME
		setuptool.setImportContext(importcontext)
		#setuptool.runAllImportSteps() #dangerous in Plone 2.5
		setuptool.runImportStep(step_id="typeinfo")
		#setuptool.runImportStep(step_id="factorytool")
		setuptool.runImportStep(step_id="toolset")
		setuptool.runImportStep(step_id="rolemap")
		setuptool.runImportStep(step_id="workflow")
		setuptool.runImportStep(step_id="skins")
		setuptool.runImportStep(step_id="cssregistry")
		setuptool.runImportStep(step_id="controlpanel")
		setuptool.runImportStep(step_id="action-icons")
		setuptool.runImportStep(step_id="jsregistry")
		setuptool.runImportStep(step_id="content_type_registry")
		#setuptool.runImportStep(step_id="catalog_tool")		

		from Products.Eldis.setuphandlers import updateRoleMappings, postInstall
		context = setuptool._getImportContext(importcontext)
		updateRoleMappings(context)
		postInstall(context)
		print >> out, "Successfully imported steps."
		
		catalogTool = getToolByName(self, 'portal_catalog')
		if 'sourceId' not in catalogTool.indexes():
				catalogTool.addIndex('sourceId', 'FieldIndex')
		print >> out, "Successfully created indexes."

		# add Property sheet to portal_properies
		pp = getToolByName(self, 'portal_properties')
		if not 'ELDISAPI_PROPERTIES' in pp.objectIds():
				pp.addPropertySheet(id='ELDISAPI_PROPERTIES', title= '%s Properties' % 'ELDISAPI_PROPERTIES')
				out.write("Adding %s property sheet to portal_properies\n" % 'ELDISAPI_PROPERTIES' )

		props_sheet = pp['ELDISAPI_PROPERTIES']
		updateProperties(props_sheet, out, PROPERTIES)

		# add the configlets to the portal control panel
		configTool = getToolByName(self, 'portal_controlpanel', None)
		if configTool:
				for conf in configlets:
						configTool.registerConfiglet(**conf)
						out.write('Added configlet %s\n' % conf['id'])

		print >> out, "Successfully installed %s." % PROJECTNAME
		return out.getvalue()
		
def uninstall(self):
		out = StringIO()

		# remove the configlets from the portal control panel
		configTool = getToolByName(self, 'portal_controlpanel', None)
		if configTool:
				for conf in configlets:
						configTool.unregisterConfiglet(conf['id'])
						out.write('Removed configlet %s\n' % conf['id'])

		try:
				self.portal_controlpanel.unregisterApplication(PROJECTNAME)
		except:
				pass

		print >> out, "Successfully uninstalled %s." % PROJECTNAME
		return out.getvalue()	

def updateProperties(pp_ps, out, *args):
		for prop in args:
				for prop_id, prop_value, prop_type in prop:
						if not pp_ps.hasProperty(prop_id):
								pp_ps.manage_addProperty(prop_id, prop_value, prop_type)
								out.write("Adding %s property to %s property sheet\n" % (prop_id, 'ELDISAPI_PROPERTIES') )


Extensions/get_eldis_rss.py
# -*- coding: utf-8 -*-
#
# File: get_eldis_rss.py
#
__version__ = '$Id: get_eldis_rss.py 159 2012-05-29 08:26:46Z subhendu $'
__author__ = """Subhendu Kumar Giri <subhendu.giri@oneworld.net>"""
__docformat__ = 'plaintext'

import urllib2
#from standards import url_quote
from xml.dom import minidom, Node
import gc
gc.enable()
""" Get the XML """

def get_eldis_article(xml_host_url,xml_info):
    #xml_url = xml_host_url+url_quote(xml_info)
    xml_url = xml_host_url+xml_info	
    url_info = urllib2.urlopen(xml_url)
    del xml_url

    if (url_info):
        xmldoc = minidom.parse(url_info)
        eldis_data=[]
        if (xmldoc):
            rootNode = xmldoc.documentElement
            resultNode = rootNode.childNodes[0]
            metaNode = rootNode.childNodes[1]
            #Iterate the child nodes
            for node in resultNode.childNodes:
                headline=""
                title=""
                if (node.nodeName == "list-item"):
                    dict_xml={}
                    #Now iterate through all of the <list-item>s children
                    for item_node in node.childNodes:
                        if (item_node.nodeName == "object_id"):
                            object_id = ""
                            for text_node in item_node.childNodes:
                                if (text_node.nodeType == node.TEXT_NODE):
                                    object_id += text_node.nodeValue
                            dict_xml['object_id'] = object_id
  
                        if (item_node.nodeName == "title"):
                            title= ""
                            for text_node in item_node.childNodes:
                                if (text_node.nodeType == node.TEXT_NODE):
                                    title+= text_node.nodeValue
                            dict_xml['title'] = title
                            
                        if (item_node.nodeName == "urls"):
                            listing = []
                            for list_node in item_node.childNodes:
                                if (list_node.nodeName == "list-item"):
                                    #Now iterate through all of the <list-item>s children
                                    for list_item in list_node.childNodes:
                                        if (list_item.nodeType == node.TEXT_NODE):
                                            listing.append('<li><a target="_blank" class="external" href="'+list_item.nodeValue+'">'+list_item.nodeValue+'</a></li>')
                            dict_xml['listing'] = '<ul>'+"".join(listing)+'</ul>'

                        if (item_node.nodeName == "category_theme_array"):
                            keyword = []
                            for list_node in item_node.childNodes: # theme
                                for node_list in list_node.childNodes: # list-item
                                    for item_list in node_list.childNodes: # object_name
                                        if (item_list.nodeName == "object_name"):
                                            for list_item in item_list.childNodes:
                                                if (list_item.nodeType == node.TEXT_NODE):
                                                    keyword.append(list_item.nodeValue)
                            dict_xml['keyword'] = ", ".join(keyword)
                            
                        if (item_node.nodeName == "website_url"):
                            website_url = ""
                            for text_node in item_node.childNodes:
                                if (text_node.nodeType == node.TEXT_NODE):
                                    website_url += text_node.nodeValue
                            dict_xml['website_url'] = website_url

                        if (item_node.nodeName == "description"):
                            long_abstract = ""
                            for text_node in item_node.childNodes:
                                if (text_node.nodeType == node.TEXT_NODE):
                                    long_abstract += text_node.nodeValue
                            dict_xml['description'] = long_abstract

                    eldis_data.append(dict_xml)
                    del dict_xml

        else:
            return "Error getting XML document!"
    else:
        return "Error! Getting URL"
    del xmldoc
    return eldis_data

config.py
# -*- coding: utf-8 -*-
#
# File: Eldis.py
#
#
# GNU General Public License (GPL)
#
__version__ = '$Id: config.py 159 2012-05-29 08:26:46Z subhendu $'
__author__ = """Subhendu Kumar Giri <subhendu.giri@oneworld.net>"""
__docformat__ = 'plaintext'

from Products.CMFCore.permissions import setDefaultRoles

PROJECTNAME = "Eldis"

# Permissions
DEFAULT_ADD_CONTENT_PERMISSION = "Add portal content"
setDefaultRoles(DEFAULT_ADD_CONTENT_PERMISSION, ('Manager', 'Owner', 'Contributor'))
ADD_CONTENT_PERMISSIONS = {
    'EldisArticle': 'Eldis: Add EldisArticle',
}

setDefaultRoles('Eldis: Add EldisArticle', ('Manager','Owner'))

product_globals = globals()

# Dependencies of Products to be installed by quick-installer
# override in custom configuration
DEPENDENCIES = []

# Dependend products - not quick-installed - used in testcase
# override in custom configuration
PRODUCT_DEPENDENCIES = []


eldis_states = (('publish','submit','draft'))
PROPERTIES = (('ELDIS_GUID', 'xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxxx-xx', 'string'),
              ('ELDIS_HOST', 'http://api.ids.ac.uk/', 'string'),
              ('ELDIS_FORMAT', 'full', 'string'),							
              ('ELDIS_DATATYPE', '', 'string'),
              ('ELDIS_NO_RECORDS', '20', 'string'),							
              ('ELDIS_START_RECORDS', '0', 'string'),								
              ('ELDIS_PUBLISHED_AFTER', '', 'string'),
              ('ELDIS_PUBLISHED_BEFORE', '', 'string'),	
              ('ELDIS_PUBLISHED_YEAR', '', 'string'),
              ('ELDIS_FOLDER', 'ids-document', 'string'),
              ('ELDIS_OBJECT', 'documents', 'string'),
              ('ELDIS_STATE', 'submit', 'string'),
              ('ELDIS_AUTHOR', '', 'string'),
              ('ELDIS_PUBLISHER', '', 'string'),
              ('ELDIS_LANGUAGE', '', 'string'),
              ('ELDIS_COUNTRY', '', 'string'),
              ('ELDIS_KEYWORD', '', 'string'),
              ('ELDIS_THEME', '', 'string'))

try:
    from Products.Eldis.AppConfig import *
except ImportError:
    pass
